enterprise/coderd/prebuilds/claim_test.go

```diff
@@ -0,0 +1,367 @@
+package prebuilds_test
+
+import (
+	"context"
+	"database/sql"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+
+	"github.com/coder/quartz"
+
+	"github.com/coder/coder/v2/coderd/coderdtest"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/dbtestutil"
+	agplprebuilds "github.com/coder/coder/v2/coderd/prebuilds"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/enterprise/coderd/coderdenttest"
+	"github.com/coder/coder/v2/enterprise/coderd/prebuilds"
+	"github.com/coder/coder/v2/provisioner/echo"
+	"github.com/coder/coder/v2/provisionersdk/proto"
+	"github.com/coder/coder/v2/testutil"
+)
+
+type storeSpy struct {
+	database.Store
+
+	claims           *atomic.Int32
+	claimParams      *atomic.Pointer[database.ClaimPrebuiltWorkspaceParams]
+	claimedWorkspace *atomic.Pointer[database.ClaimPrebuiltWorkspaceRow]
+}
+
+func newStoreSpy(db database.Store) *storeSpy {
+	return &storeSpy{
+		Store:            db,
+		claims:           &atomic.Int32{},
+		claimParams:      &atomic.Pointer[database.ClaimPrebuiltWorkspaceParams]{},
+		claimedWorkspace: &atomic.Pointer[database.ClaimPrebuiltWorkspaceRow]{},
+	}
+}
+
+func (m *storeSpy) InTx(fn func(store database.Store) error, opts *database.TxOptions) error {
+	// Pass spy down into transaction store.
+	return m.Store.InTx(func(store database.Store) error {
+		spy := newStoreSpy(store)
+		spy.claims = m.claims
+		spy.claimParams = m.claimParams
+		spy.claimedWorkspace = m.claimedWorkspace
+
+		return fn(spy)
+	}, opts)
+}
+
+func (m *storeSpy) ClaimPrebuiltWorkspace(ctx context.Context, arg database.ClaimPrebuiltWorkspaceParams) (database.ClaimPrebuiltWorkspaceRow, error) {
+	m.claims.Add(1)
+	m.claimParams.Store(&arg)
+	result, err := m.Store.ClaimPrebuiltWorkspace(ctx, arg)
+	if err == nil {
+		m.claimedWorkspace.Store(&result)
+	}
+	return result, err
+}
+
+func TestClaimPrebuild(t *testing.T) {
+	t.Parallel()
+
+	if !dbtestutil.WillUsePostgres() {
+		t.Skip("This test requires postgres")
+	}
+
+	const (
+		desiredInstances = 1
+		presetCount      = 2
+	)
+
+	cases := map[string]struct {
+		expectPrebuildClaimed  bool
+		markPrebuildsClaimable bool
+	}{
+		"no eligible prebuilds to claim": {
+			expectPrebuildClaimed:  false,
+			markPrebuildsClaimable: false,
+		},
+		"claiming an eligible prebuild should succeed": {
+			expectPrebuildClaimed:  true,
+			markPrebuildsClaimable: true,
+		},
+		"claiming an eligible prebuild results in error": {
+			expectPrebuildClaimed:  true,
+			markPrebuildsClaimable: true,
+		},
```

Comment:
These two test cases are identical

Reply:
Whoops, that's something I forgot to revert; I was adding an error case but rolled it back (incompletely, it appears) because it would've been too big a divergence from your branch.

Reply:
okay, I'll remove it for now